### PR TITLE
Add SCC operations to VsAwareFileSystem

### LINKED ISSUE #4387: Implement full IVsAwareFileSystem
### Description

The current implementation is not VS-aware, and just passes 
calls through the file system.

Three separate areas, in priority order:
* source-code control - notify VS source-controlled files are being modified. Important for a minority of users who are not using git for source control. The old binding method correctly handled this situation.
* reading file contents from in-memory: this would prevent conflicts/prompts when VS notices we've changed the file on disc outside the solution. However, the migration guidance to users is to start from a clean slate before migrating i.e. no unsaved files, and no outstanding source-control changes.
* notify VS that we have changed a file outside the solution so it does not prompt the user.

### PR DESCRIPTION
**Pull request by duncanp-sonar**
Fixes #4387

### CODE REVIEW

File: src/ConnectedMode/Migration/VsAwareFileSystem.cs [hunk 1]
```diff
@@ -58,52 +61,148 @@ internal class VsAwareFileSystem : IVsAwareFileSystem
...
+            }
+
+            if (!success)
+            {
+                var message = string.Format(MigrationStrings.VsFileSystem_Error_FailedToCheckOutFolderForDeletion, folderPath);
+                throw new InvalidOperationException(message);
+            }
+        }
+
+        private async Task<bool> CheckoutForEditAsync(params string[] fileNames)
```
Comment (duncanp-sonar) [inline src/ConnectedMode/Migration/VsAwareFileSystem.cs:148]: This was copied from the [equivalent method](https://github.com/SonarSource/sonarlint-visualstudio/blob/03123f51a5f22aa2bd6e347f00a33ca01e9ee44b/src/Integration/LocalServices/SourceControlledFileSystem.cs#L133) of the existing `SourceControlledFileSystem` class.

I've tweaked it slightly to make it `async` so we can force the SCC operation to happen on the UI thread (VS crashes if it is called on a background thread),

The legacy `SourceControlledFileSystem` implementation assumed that the caller would ensure that it was called on the main thread.

File: src/ConnectedMode/Migration/VsAwareFileSystem.cs [hunk 2]
```diff
@@ -58,52 +61,148 @@ internal class VsAwareFileSystem : IVsAwareFileSystem
...
+        {
+            bool result = false;
+
+            await threadHandling.RunOnUIThread(() =>
+                result = DoCheckoutForSave(fileNames));
+
+            return result;
+        }
+
+        private bool DoCheckoutForSave(params string[] fileNames)
```
Comment (duncanp-sonar) [inline src/ConnectedMode/Migration/VsAwareFileSystem.cs:185]: Copied the existing [method](https://github.com/SonarSource/sonarlint-visualstudio/blob/03123f51a5f22aa2bd6e347f00a33ca01e9ee44b/src/Integration/LocalServices/SourceControlledFileSystem.cs#L158).

I didn't change this method to handle threading; it was simpler to add a thread-aware wrapper method instead.

File: src/ConnectedMode/Migration/VsAwareFileSystem.cs [hunk 3]
```diff
@@ -58,52 +61,148 @@ internal class VsAwareFileSystem : IVsAwareFileSystem
...
 
-        public Task SaveAsync(string filePath, string text)
+        public async Task SaveAsync(string filePath, string text)
         {
-            // TODO - error handling
+            Debug.Assert(batchStarted, "Expecting the changes to happen in a batch");
+
             logger.LogMigrationVerbose($"Saving file: {filePath}");
+
+            bool success = await CheckoutForEditAsync(filePath);
```
Comment (duncanp-sonar) [inline src/ConnectedMode/Migration/VsAwareFileSystem.cs:88]: Before we save a file we need to tell VS we are going to check it out.
If we can't check it out we will throw, which will cause the migration to end.

File: src/ConnectedMode/Migration/VsAwareFileSystem.cs [hunk 4]
```diff
@@ -58,52 +61,148 @@ internal class VsAwareFileSystem : IVsAwareFileSystem
...
 
         public async Task BeginChangeBatchAsync()
         {
+            Debug.Assert(!batchStarted, "Not expecting to already be in a batch");
+
             logger.LogMigrationVerbose("Beginning batch of file changes...");
 
-            queryFileOperation = await GetSccServiceAsync();
+            sccService = await GetSccServiceAsync();
+            sccService.BeginQuerySaveBatch();
```
Comment (duncanp-sonar) [inline src/ConnectedMode/Migration/VsAwareFileSystem.cs:106]: Simple pass-through to the VS API.

File: src/ConnectedMode/Migration/VsAwareFileSystem.cs [hunk 5]
```diff
@@ -58,52 +61,148 @@ internal class VsAwareFileSystem : IVsAwareFileSystem
...
             });
 
             return sccService;
         }
+
+        private async Task CheckOutFilesToDeleteAsync(string folderPath)
+        {
+            // Can't check out the directory for saving - if you do, VS will pop up a dialogue,
+            // even if the "silent" flag is set.
+            bool success = await CheckoutForEditAsync(folderPath);
```
Comment (duncanp-sonar) [inline src/ConnectedMode/Migration/VsAwareFileSystem.cs:133]: Tell VS we are going to "change" the directory...

File: src/ConnectedMode/Migration/VsAwareFileSystem.cs [hunk 6]
```diff
@@ -58,52 +61,148 @@ internal class VsAwareFileSystem : IVsAwareFileSystem
...
+
+            if (success)
+            {
+                var files = fileSystem.Directory.GetFiles(folderPath, "*.*", System.IO.SearchOption.AllDirectories);
+                success = await CheckoutForSaveAsync(files);
```
Comment (duncanp-sonar) [inline src/ConnectedMode/Migration/VsAwareFileSystem.cs:138]: Telling VS that we are going to edit/save the `.sonarlint` folder doesn't appear to be enough on its own - we have to tell VS we are going to change each file under the folder.

If we don't, the deleted files don't appear as `Pending changes` in the `Solution Explorer`.

File: src/ConnectedMode/Migration/ConnectedModeMigration.cs [hunk 7]
```diff
@@ -136,7 +136,7 @@ private async Task MigrateImplAsync(BoundSonarQubeProject oldBinding, IProgress<
 
             // Now make all of the files changes required to remove the legacy settings
             // i.e. update project files and delete .sonarlint folder
-            await MakeLegacyFileChangesAsync(legacySettings, changedFiles, progress);
+            await MakeLegacyFileChangesAsync(legacySettings, changedFiles, progress, token);
```
Comment (duncanp-sonar) [inline src/ConnectedMode/Migration/ConnectedModeMigration.cs:139]: Passing the token down so we can check for cancellation.

**Comment by duncanp-sonar** [comment]
I've already merged the first four commits since they were simple refactoring.
This PR covers the last two - adding the SCC operations to the VSAwareFileSystem class.

**Comment by duncanp-sonar** [comment]
I'll fix the merge conflicts tomorrow.

File: src/ConnectedMode.UnitTests/Migration/VsAwareFileSystemTests.cs [hunk 8]
```diff
@@ -18,11 +18,17 @@
  * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
  */
```
Comment (bigfluffycookie) [inline src/ConnectedMode.UnitTests/Migration/VsAwareFileSystemTests.cs:20]: We use threadhandling 3 times in the actual class and I think we should have some tests covering that it runs on the expected thread

**Review by bigfluffycookie — COMMENTED** [review]
Looks good but maybe some missing test cases

File: src/ConnectedMode.UnitTests/Migration/VsAwareFileSystemTests.cs (hunk 8 above)
Comment (duncanp-sonar) [reply src/ConnectedMode.UnitTests/Migration/VsAwareFileSystemTests.cs:20]: @bigfluffycookie I've added threading unit tests for the two cases where we use the VS interface (the calls that crash VS if we are not on the UI thread).

Comment (duncanp-sonar) [reply src/ConnectedMode.UnitTests/Migration/VsAwareFileSystemTests.cs:20]: (and I fixed the code smell)

**Pull request merged**